File: examples/src/test/scala/docs/state/DurableStateChangeEventDocExample.scala
/*
 * Copyright (C) 2023-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.state

import scala.annotation.nowarn
import scala.concurrent.duration._

import akka.Done
import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.actor.typed.SupervisorStrategy
import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
import akka.persistence.typed.PersistenceId

// #changeEventHandler
import akka.persistence.typed.state.scaladsl.ChangeEventHandler
import akka.persistence.typed.state.scaladsl.DurableStateBehavior
import akka.persistence.typed.state.scaladsl.Effect

// #changeEventHandler
import akka.serialization.jackson.CborSerializable

object DurableStateChangeEventDocExample {
  // #changeEventHandler
  object ShoppingCart {

    // #changeEventHandler

    final case class State(items: Map[String, Int]) extends CborSerializable {

      def hasItem(itemId: String): Boolean =
        items.contains(itemId)

      def isEmpty: Boolean =
        items.isEmpty

      def updateItem(itemId: String, quantity: Int): State = {
        quantity match {
          case 0 => copy(items = items - itemId)
          case _ => copy(items = items + (itemId -> quantity))
        }
      }

      def removeItem(itemId: String): State =
        copy(items = items - itemId)
    }
    object State {
      val empty = State(items = Map.empty)
    }

    sealed trait Command extends CborSerializable

    final case class AddItem(itemId: String, quantity: Int, replyTo: ActorRef[Done]) extends Command
    final case class RemoveItem(itemId: String, replyTo: ActorRef[Done]) extends Command
    final case class DiscardCart(replyTo: ActorRef[Done]) extends Command

    sealed trait ChangeEvent extends CborSerializable
    final case class ItemAdded(itemId: String, quantity: Int) extends ChangeEvent
    final case class ItemRemoved(itemId: String) extends ChangeEvent
    case object CartRemoved extends ChangeEvent

    val EntityKey: EntityTypeKey[Command] = EntityTypeKey[Command]("ShoppingCart")

    @nowarn("msg=never used")
    // #changeEventHandler
    val changeEventHandler = ChangeEventHandler[Command, State, ChangeEvent](
      updateHandler = {
        case (previousState, newState, AddItem(itemId, quantity, _)) =>
          ItemAdded(itemId, quantity)
        case (previousState, newState, RemoveItem(itemId, _)) =>
          ItemRemoved(itemId)
        case (_, _, command @ DiscardCart(_)) =>
          throw new IllegalStateException(s"Unexpected command ${command.getClass}")
      },
      deleteHandler = { (previousState, command) =>
        CartRemoved
      })

    def apply(cartId: String): Behavior[Command] = {
      DurableStateBehavior
        .withEnforcedReplies[Command, State](
          PersistenceId(EntityKey.name, cartId),
          State.empty,
          (state, command) =>
            command match {
              case AddItem(itemId, quantity, replyTo) =>
                val newState = state.updateItem(itemId, quantity)
                Effect.persist(newState).thenReply(replyTo)(_ => Done)
              case RemoveItem(itemId, replyTo) =>
                val newState = state.removeItem(itemId)
                Effect.persist(newState).thenReply(replyTo)(_ => Done)
              case DiscardCart(replyTo) =>
                Effect.delete().thenReply(replyTo)(_ => Done)
            })
        .withChangeEventHandler(changeEventHandler)
        .onPersistFailure(SupervisorStrategy.restartWithBackoff(200.millis, 5.seconds, 0.1))
    }

  }
  // #changeEventHandler

}

File: examples/src/test/java/jdocs/state/DurableStateChangeEventDocExample.java
/*
 * Copyright (C) 2023-2025 Lightbend Inc. <https://akka.io>
 */

package jdocs.state;

import akka.Done;
import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.SupervisorStrategy;
import akka.cluster.sharding.typed.javadsl.EntityTypeKey;
import akka.persistence.typed.PersistenceId;
// #changeEventHandler
import akka.persistence.typed.state.javadsl.ChangeEventHandler;
import akka.persistence.typed.state.javadsl.CommandHandlerWithReply;
import akka.persistence.typed.state.javadsl.CommandHandlerWithReplyBuilder;
import akka.persistence.typed.state.javadsl.DurableStateBehaviorWithEnforcedReplies;
import akka.persistence.typed.state.javadsl.ReplyEffect;

// #changeEventHandler
import akka.serialization.jackson.CborSerializable;
import com.fasterxml.jackson.annotation.JsonCreator;

import java.time.Duration;
import java.util.HashMap;
import java.util.Map;

public class DurableStateChangeEventDocExample {
  static
  // #changeEventHandler
  public class ShoppingCart
      extends DurableStateBehaviorWithEnforcedReplies<ShoppingCart.Command, ShoppingCart.State> {
    // #changeEventHandler

    public final static class State implements CborSerializable {
      private Map<String, Integer> items = new HashMap<>();

      public boolean isEmpty() {
        return items.isEmpty();
      }

      public boolean hasItem(String itemId) {
        return items.containsKey(itemId);
      }

      public State updateItem(String itemId, int quantity) {
        if (quantity == 0) {
          items.remove(itemId);
        } else {
          items.put(itemId, quantity);
        }
        return this;
      }

      public State removeItem(String itemId) {
        items.remove(itemId);
        return this;
      }

    }

    public interface Command extends CborSerializable {
    }

    public static class AddItem implements Command {
      public final String itemId;
      public final int quantity;
      public final ActorRef<Done> replyTo;

      public AddItem(String itemId, int quantity, ActorRef<Done> replyTo) {
        this.itemId = itemId;
        this.quantity = quantity;
        this.replyTo = replyTo;
      }
    }

    public static class RemoveItem implements Command {
      public final String itemId;
      public final ActorRef<Done> replyTo;

      public RemoveItem(String itemId, ActorRef<Done> replyTo) {
        this.itemId = itemId;
        this.replyTo = replyTo;
      }
    }

    public static class DiscardCart implements Command {
      public final ActorRef<Done> replyTo;

      @JsonCreator
      public DiscardCart(ActorRef<Done> replyTo) {
        this.replyTo = replyTo;
      }
    }

    public interface ChangeEvent extends CborSerializable {
    }

    public static final class ItemAdded implements ChangeEvent {
      public final String itemId;
      public final int quantity;

      public ItemAdded(String itemId, int quantity) {
        this.itemId = itemId;
        this.quantity = quantity;
      }
    }

    public static final class ItemRemoved implements ChangeEvent {
      public final String itemId;

      @JsonCreator
      public ItemRemoved(String itemId) {
        this.itemId = itemId;
      }
    }

    public static final class CartRemoved implements ChangeEvent {
    }

    public static EntityTypeKey<Command> ENTITY_TYPE_KEY =
        EntityTypeKey.create(Command.class, "ShoppingCart");

    public static Behavior<Command> create(String cartId) {
      return new ShoppingCart(cartId);
    }

    private final String cartId;

    private ShoppingCart(String cartId) {
      super(
          PersistenceId.of(ENTITY_TYPE_KEY.name(), cartId),
          SupervisorStrategy.restartWithBackoff(Duration.ofMillis(200), Duration.ofSeconds(5), 0.1));
      this.cartId = cartId;
    }

    @Override
    public State emptyState() {
      return new State();
    }

    // #changeEventHandler

    @Override
    public ChangeEventHandler<Command, State, ChangeEvent> changeEventHandler() {
      return new ChangeEventHandler<Command, State, ChangeEvent>() {
        @Override
        public ChangeEvent updateHandler(State previousState, State newState, Command command) {
          if (command instanceof AddItem) {
            AddItem addItem = (AddItem) command;
            return new ItemAdded(addItem.itemId, addItem.quantity);
          } else if (command instanceof RemoveItem) {
            RemoveItem removeItem = (RemoveItem) command;
            return new ItemRemoved(removeItem.itemId);
          } else {
            throw new IllegalStateException("Unexpected command " + command.getClass());
          }
        }

        @Override
        public ChangeEvent deleteHandler(State previousState, Command command) {
          return new CartRemoved();
        }
      };
    }

    @Override
    public CommandHandlerWithReply<Command, State> commandHandler() {
      CommandHandlerWithReplyBuilder<Command, State> b = newCommandHandlerWithReplyBuilder();

      b.forAnyState()
          .onCommand(AddItem.class, this::onAddItem)
          .onCommand(RemoveItem.class, this::onRemoveItem)
          .onCommand(DiscardCart.class, this::onDiscardCart);

      return b.build();
    }

    public ReplyEffect<State> onAddItem(State state, AddItem cmd) {
      return Effect()
          .persist(state.updateItem(cmd.itemId, cmd.quantity))
          .thenReply(cmd.replyTo, updatedCart -> Done.getInstance());
    }

    public ReplyEffect<State> onRemoveItem(State state, RemoveItem cmd) {
      return Effect()
          .persist(state.removeItem(cmd.itemId))
          .thenReply(cmd.replyTo, updatedCart -> Done.getInstance());
    }

    public ReplyEffect<State> onDiscardCart(State state, DiscardCart cmd) {
      return Effect()
          .delete()
          .thenReply(cmd.replyTo, updatedCart -> Done.getInstance());
    }

  }
  // #changeEventHandler
}
